scripts/mkimg.sh

```diff
@@ -9,23 +9,64 @@
 # The golden image should be as small as possible while still achieving good
 # coverage, so as to minimize the size of data stored in git.
 
-truncate -s 64m resources/ufs.img
-MD=$(mdconfig -a -t vnode -f resources/ufs.img)
-newfs $MD
-MNTDIR=`mktemp -d`
-mount -t ufs /dev/"$MD" "$MNTDIR"
-
-# TODO: create files and directories
-cd "$MNTDIR"
-echo 'This is a simple file.' > file1
-mkdir -p dir1/dir2/dir3
-echo 'Hello World' > dir1/dir2/dir3/file2
-jot $((1 << 16)) 0 | xargs printf '%015x\n' > file3
-ln -sf dir1/dir2/dir3/file2 link1
-cd -
-
-umount "$MNTDIR"
-rmdir "$MNTDIR"
-mdconfig -d -u "$MD"
-
-zstd -f resources/ufs.img
+die() {
+    echo "ERROR: $*" >&2
+    exit 1
+}
+
+# $1: mountpoint
+populate() {
+    cd "$1" || die "failed to cd into '$1'"
+
+    echo 'This is a simple file.' > file1
+    mkdir -p dir1/dir2/dir3
+    echo 'Hello World' > dir1/dir2/dir3/file2
+    jot $((1 << 16)) 0 | xargs printf '%015x\n' > file3
+    ln -sf dir1/dir2/dir3/file2 link1
+    ln -sf "$(yes ./ | head -n200 | tr -d '\n')/file1" long-link
+
+    cd - || die "failed to cd back"
+}
+
+# $1: name
+# $2: size
+# $@: args to newfs
+create() {
+    name=$1
+    path=resources/${name}.img
+    size=$2
+    shift 2
+    
+    truncate -s "$size" "$path" || die "$path: failed to allocate $size"
+    dev=$(mdconfig -a -t vnode -f "$path") || die "$path: failed to create virtual device"
+    newfs "$@" "$dev" || die "$path: failed to newfs $dev"
+
+    mnt=$(mktemp -d) || die "$path: failed to create tempdir"
+    mount -t ufs "/dev/$dev" "$mnt" || die "$path: failed to mount '/dev/$dev' onto '$mnt'"
+
+    populate "$mnt"
+
+    # These may fail with only a warning:
+    umount "$mnt"
+    rmdir "$mnt"
+    mdconfig -d -u "$dev"
+
+    zstd -f "$path" || die "$path: failed to compress with zstd"
+}
+
+# I don't know why it works, but it does. Tested on FreeBSD/powerpc64, OpenBSD/amd64, Arch Linux (amd64)
```

Comment:
Interesting trick.  Where did you learn it?

Reply:
Unfortunately I didn't come up with it. I adapted it from StackOverflow.